provider/datagen/src/lib.rs

```diff
@@ -177,73 +183,166 @@ impl DatagenProvider {
+                        if key
+                            == icu_datetime::provider::calendar::DateSkeletonPatternsV1Marker::KEY
+                            && !locale.has_unicode_ext()
+                        {
+                            return false;
+                        }
+                        let mut iter = fallbacker_with_config.fallback_for(locale.clone());
+                        while !iter.get().is_empty() {
+                            if explicit.contains(iter.get()) {
                                 return true;
                             }
-                            chain.step();
+                            iter.step();
                         }
+                        log::trace!("Filtered out: {key}/{locale}");
                         false
-                    } else {
-                        set.contains(&l.get_langid().language.into())
-                    }
-                })
-                .collect(),
-            _ => unreachable!("resolved"),
+                    })
+                    .collect()
+            }
+            _ => unreachable!("Pre-processed LocaleInclude has only 2 variants"),
         };
 
-        Ok(
-            if key == icu_collator::provider::CollationDataV1Marker::KEY
-                || key == icu_collator::provider::CollationDiacriticsV1Marker::KEY
-                || key == icu_collator::provider::CollationJamoV1Marker::KEY
-                || key == icu_collator::provider::CollationMetadataV1Marker::KEY
-                || key == icu_collator::provider::CollationReorderingV1Marker::KEY
-                || key == icu_collator::provider::CollationSpecialPrimariesV1Marker::KEY
-            {
-                transform::icuexport::collator::filter_data_locales(
-                    resolved_locales,
-                    &options.collations,
-                )
-            } else if key == icu_segmenter::provider::LstmForWordLineAutoV1Marker::KEY {
-                transform::segmenter::lstm::filter_data_locales(
-                    resolved_locales,
-                    &options.segmenter_models,
-                )
-            } else if key == icu_segmenter::provider::DictionaryForWordOnlyAutoV1Marker::KEY
-                || key == icu_segmenter::provider::DictionaryForWordLineExtendedV1Marker::KEY
-            {
-                transform::segmenter::dictionary::filter_data_locales(
-                    resolved_locales,
-                    &options.segmenter_models,
-                )
-            } else {
-                resolved_locales
-            },
-        )
+        Ok(locales)
+    }
+
+    /// Loads a `DataPayload<ExportMarker>` with locale fallback enabled.
+    fn load_with_fallback(
+        &self,
+        key: DataKey,
+        locale: &DataLocale,
+        fallbacker: &LocaleFallbacker,
+    ) -> Result<Option<DataPayload<ExportMarker>>, DataError> {
+        log::trace!("Generating key/locale: {key}/{locale:}");
```

Comment:
nit: not a big fan of inlining the whole fallback algorithm just to add two debug statements.